orttraining/orttraining/core/optimizer/bias_softmax_dropout_fusion.cc

```diff
@@ -0,0 +1,125 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "orttraining/core/optimizer/bias_softmax_dropout_fusion.h"
+
+#include "core/graph/graph_utils.h"
+#include "core/optimizer/initializer.h"
+
+namespace onnxruntime {
+
+// Both BiasSoftmax and BiasSoftmaxDropout follow the axis attribute definition from Softmax Op before OpSet-11,
+// that it, no extra transpose if axis is not rank-1. So if BiasSoftmax is found in the graph, it's safe to
+// fuse matched sub-graph to BiasSoftmaxDropout without checking the axis attribute.
+Status BiasSoftmaxDropoutFusion::ApplyImpl(Graph& graph, bool& modified, int graph_level,
+                                           const logging::Logger& logger) const {
+  GraphViewer graph_viewer(graph);
+  const auto& node_topology_list = graph_viewer.GetNodesInTopologicalOrder();
+
+  for (auto node_index : node_topology_list) {
+    auto* node_ptr = graph.GetNode(node_index);
+    if (!node_ptr) continue;  // Node was removed.
+
+    auto& node = *node_ptr;
+    ORT_RETURN_IF_ERROR(Recurse(node, modified, graph_level, logger));
+
+    // Matching for BiasSoftmax node.
+    if (!graph_utils::IsSupportedOptypeVersionAndDomain(node, "BiasSoftmax", {1}, kMSDomain) ||
```

Comment:
>BiasSoftmax [](http://example.com/codeflow?start=63&length=11)

softmax-11 + dropout can also be fused in the same kernel, I think you probably already had some thinking around that.

Reply:
I don't think it worth effort to support softmax-11 as its definition is out of date. It's possible we can fuse softmax-13 with axis=-1, but this requires extra checking in the fusion code, and also need a new Op SoftmaxDropout, or at least change the bias input optional and change the kernel code accordingly. This is doable but requires extra effort. But I think we can leave it as a model driven task, if we find such case in future model, I can add this support in the future.